File: frontend/src/api/mock/seed.ts
/**
 * A plausible evening's service, so the prototype opens on a working door
 * rather than an empty screen — and so the estimator has real turnaround
 * history to blend against its priors from the first click.
 *
 * The backend equivalent of this is `make seed`.
 */
import type { Config } from '../types';
import type { MockState, StoredParty } from './store';
import { newToken } from './token';

interface Closed {
  name: string;
  size: number;
  /** Minutes before now that the party joined the list. */
  joinedAgo: number;
  /** Minutes from joining to being seated. */
  turnaround: number;
  /** What the estimator told them at the door. */
  quoted: number;
}

const SEATED: Closed[] = [
  { name: 'Marsh', size: 2, joinedAgo: 205, turnaround: 22, quoted: 25 },
  { name: 'Ferreira', size: 4, joinedAgo: 198, turnaround: 44, quoted: 40 },
  { name: 'Nakamura', size: 2, joinedAgo: 190, turnaround: 26, quoted: 25 },
  { name: 'Oyelaran', size: 5, joinedAgo: 182, turnaround: 58, quoted: 55 },
  { name: 'Baptiste', size: 3, joinedAgo: 176, turnaround: 38, quoted: 40 },
  { name: 'Kowalczyk', size: 2, joinedAgo: 168, turnaround: 19, quoted: 25 },
  { name: 'Ainsley', size: 4, joinedAgo: 160, turnaround: 51, quoted: 40 },
  { name: 'Rasmussen', size: 7, joinedAgo: 150, turnaround: 78, quoted: 75 },
  { name: 'Delacroix', size: 2, joinedAgo: 141, turnaround: 24, quoted: 25 },
  { name: 'Haddad', size: 6, joinedAgo: 132, turnaround: 61, quoted: 55 },
  { name: 'Strand', size: 4, joinedAgo: 120, turnaround: 35, quoted: 40 },
  { name: 'Iqbal', size: 2, joinedAgo: 108, turnaround: 31, quoted: 25 },
  { name: 'Bellweather', size: 3, joinedAgo: 96, turnaround: 42, quoted: 40 },
  { name: 'Sorrentino', size: 2, joinedAgo: 84, turnaround: 23, quoted: 25 },
  { name: 'Achebe', size: 5, joinedAgo: 70, turnaround: 54, quoted: 55 },
  { name: 'Mendelsohn', size: 4, joinedAgo: 58, turnaround: 46, quoted: 40 },
];

interface Waiting {
  name: string;
  size: number;
  joinedAgo: number;
  /** Minutes before now that the host called them, if they have been called. */
  notifiedAgo?: number;
  phone?: string;
  note?: string;
  quoted: number;
}

const ON_THE_LIST: Waiting[] = [
  { name: 'Dana Whitcomb', size: 4, joinedAgo: 34, note: 'Highchair', quoted: 40 },
  { name: 'Okonjo', size: 2, joinedAgo: 27, notifiedAgo: 4, phone: '555 0147', quoted: 25 },
  { name: 'Vasquez', size: 6, joinedAgo: 22, note: 'Patio if one opens', quoted: 55 },
  { name: 'Lindqvist', size: 2, joinedAgo: 15, quoted: 25 },
  { name: 'Behrouzi', size: 8, joinedAgo: 11, note: 'Birthday, no candles', quoted: 75 },
  { name: 'Dupont', size: 3, joinedAgo: 5, phone: '555 0192', quoted: 40 },
];

export function buildSeed(config: Config): MockState {
  const now = Date.now();
  const at = (minutesAgo: number) => new Date(now - minutesAgo * 60_000).toISOString();

  const parties: StoredParty[] = [];
  let id = 1;

  const base = (name: string, size: number, joinedAgo: number, quoted: number): StoredParty => ({
    id: id++,
    token: newToken(),
    name,
    size,
    phone: null,
    note: null,
    status: 'WAITING',
    quoted_wait_minutes: quoted,
    joined_at: at(joinedAgo),
    notified_at: null,
    seated_at: null,
    closed_at: null,
  });

  for (const entry of SEATED) {
    const party = base(entry.name, entry.size, entry.joinedAgo, entry.quoted);
    const seatedAgo = entry.joinedAgo - entry.turnaround;
    party.status = 'SEATED';
    party.notified_at = at(seatedAgo + 4);
    party.seated_at = at(seatedAgo);
    party.closed_at = at(seatedAgo);
    parties.push(party);
  }

  // One party the host called who never came back, and two that gave up.
  const trelawny = base('Trelawny', 2, 92, 25);
  trelawny.status = 'NO_SHOW';
  trelawny.notified_at = at(68);
  trelawny.closed_at = at(60);
  parties.push(trelawny);

  const kaur = base('Kaur', 4, 78, 40);
  kaur.status = 'CANCELLED';
  kaur.closed_at = at(70);
  kaur.note = 'Went next door';
  parties.push(kaur);

  const novak = base('Novak', 2, 46, 25);
  novak.status = 'CANCELLED';
  novak.notified_at = at(30);
  novak.closed_at = at(26);
  parties.push(novak);

  for (const entry of ON_THE_LIST) {
    const party = base(entry.name, entry.size, entry.joinedAgo, entry.quoted);
    party.phone = entry.phone ?? null;
    party.note = entry.note ?? null;
    if (entry.notifiedAgo !== undefined) {
      party.status = 'NOTIFIED';
      party.notified_at = at(entry.notifiedAgo);
    }
    parties.push(party);
  }

  return { nextId: id, parties, config: structuredClone(config), signedIn: false };
}

File: backend/scripts/seed.py
"""Load a realistic evening so the adaptive estimator can be seen working.

Without history every quote is the configured prior, which makes the whole
point of section 6 invisible. This writes a service already in progress:
sixteen seated parties to learn from, a few that left, and six still waiting.

    uv run python scripts/seed.py [--force]
"""
import argparse
import pathlib
import sys
from datetime import UTC, datetime, timedelta

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parents[1]))

from app.db import get_session_factory  # noqa: E402
from app.models import PartyRecord, PartyStatus  # noqa: E402
from app.services.waitlist import new_token  # noqa: E402

# name, size, minutes ago they joined, minutes until seated, what we quoted
SEATED = [
    ("Marsh", 2, 205, 22, 25), ("Ferreira", 4, 198, 44, 40),
    ("Nakamura", 2, 190, 26, 25), ("Oyelaran", 5, 182, 58, 55),
    ("Baptiste", 3, 176, 38, 40), ("Kowalczyk", 2, 168, 19, 25),
    ("Ainsley", 4, 160, 51, 40), ("Rasmussen", 7, 150, 78, 75),
    ("Delacroix", 2, 141, 24, 25), ("Haddad", 6, 132, 61, 55),
    ("Strand", 4, 120, 35, 40), ("Iqbal", 2, 108, 31, 25),
    ("Bellweather", 3, 96, 42, 40), ("Sorrentino", 2, 84, 23, 25),
    ("Achebe", 5, 70, 54, 55), ("Mendelsohn", 4, 58, 46, 40),
]

# name, size, joined, called (None if never), quoted, note, phone
WAITING = [
    ("Dana Whitcomb", 4, 34, None, 40, "Highchair", None),
    ("Okonjo", 2, 27, 4, 25, None, "555 0147"),
    ("Vasquez", 6, 22, None, 55, "Patio if one opens", None),
    ("Lindqvist", 2, 15, None, 25, None, None),
    ("Behrouzi", 8, 11, None, 75, "Birthday, no candles", None),
    ("Dupont", 3, 5, None, 40, None, "555 0192"),
]


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--force", action="store_true", help="seed even if parties exist")
    args = parser.parse_args()

    now = datetime.now(UTC)
    ago = lambda minutes: now - timedelta(minutes=minutes)  # noqa: E731

    with get_session_factory()() as session:
        existing = session.query(PartyRecord).count()
        if existing and not args.force:
            print(f"{existing} parties already here. Use --force to add anyway.")
            return

        for name, size, joined, turnaround, quoted in SEATED:
            seated_at = ago(joined - turnaround)
            session.add(PartyRecord(
                token=new_token(), name=name, size=size, quoted_wait_minutes=quoted,
                joined_at=ago(joined), status=PartyStatus.SEATED,
                notified_at=ago(joined - turnaround + 4),
                seated_at=seated_at, closed_at=seated_at,
            ))

        session.add(PartyRecord(
            token=new_token(), name="Trelawny", size=2, quoted_wait_minutes=25,
            joined_at=ago(92), status=PartyStatus.NO_SHOW,
            notified_at=ago(68), closed_at=ago(60),
        ))
        session.add(PartyRecord(
            token=new_token(), name="Kaur", size=4, quoted_wait_minutes=40,
            joined_at=ago(78), status=PartyStatus.CANCELLED,
            note="Went next door", closed_at=ago(70),
        ))

        for name, size, joined, called, quoted, note, phone in WAITING:
            session.add(PartyRecord(
                token=new_token(), name=name, size=size, quoted_wait_minutes=quoted,
                joined_at=ago(joined), note=note, phone=phone,
                status=PartyStatus.NOTIFIED if called else PartyStatus.WAITING,
                notified_at=ago(called) if called else None,
            ))

        session.commit()

    print(f"Seeded {len(SEATED)} seated parties, 2 that left, and {len(WAITING)} on the list.")


if __name__ == "__main__":
    main()
